numpyro/infer/mcmc.py

```diff
@@ -733,75 +742,127 @@ def _hashable(x):
+
+    def _single_chain_mcmc(self, rng_key, init_state, init_params, args, kwargs, collect_fields=('z',)):
         num_warmup = self.num_warmup
-        if reuse_warmup_state:
-            if self._warmup_state is None:
-                raise ValueError('No `init_state` found; warmup results cannot be reused.')
-            num_warmup = 0
-            init_state = self._warmup_state._replace(rng_key=rng_key)
-        else:
+        if init_state is None:
             init_state = self.sampler.init(rng_key, self.num_warmup, init_params,
                                            model_args=args, model_kwargs=kwargs)
         if self.constrain_fn is None:
             self.constrain_fn = self.sampler.constrain_fn(args, kwargs)
-        lower = 0 if collect_warmup else num_warmup
         diagnostics = lambda x: get_diagnostics_str(x[0]) if rng_key.ndim == 1 else None   # noqa: E731
         init_val = (init_state, args, kwargs) if self._jit_model_args else (init_state,)
-        collect_vals = fori_collect(lower, num_warmup + self.num_samples,
+        collect_vals = fori_collect(self._fori_collect_infos["lower"],
+                                    self._fori_collect_infos["upper"],
                                     self._get_cached_fn(),
                                     init_val,
-                                    transform=_collect_fn(collect_fields),  # noqa: E731
+                                    transform=_collect_fn(collect_fields),
                                     progbar=self.progress_bar,
-                                    return_init_state=True,
+                                    return_last_val=True,
+                                    collect_size=self._fori_collect_infos["collect_size"],
                                     progbar_desc=functools.partial(get_progbar_desc_str, num_warmup),
                                     diagnostics_fn=diagnostics)
-        states, warmup_state = collect_vals
+        states, last_val = collect_vals
         # Get first argument of type `HMCState`
-        self._warmup_state = warmup_state[0]
+        last_state = last_val[0]
         if len(collect_fields) == 1:
             states = (states,)
         states = dict(zip(collect_fields, states))
         # Apply constraints if number of samples is non-zero
         site_values = tree_flatten(states['z'])[0]
         if len(site_values) > 0 and site_values[0].size > 0:
             states['z'] = lax.map(self.constrain_fn, states['z'])
-        return states
+        return states, last_state
+
+    def _single_chain_jit_args(self, init, collect_fields=('z',)):
+        return self._single_chain_mcmc(*init, collect_fields=collect_fields)
+
+    def _single_chain_nojit_args(self, init, model_args, model_kwargs, collect_fields=('z',)):
+        return self._single_chain_mcmc(*init, model_args, model_kwargs, collect_fields=collect_fields)
+
+    def _set_fori_collect_infos(self, lower=None, upper=None, collect_size=None):
```

Comment:
nit: can we rename this to `_set_collection_arguments(..)` or `_set_collection_params` or something? This can also be a namedtuple `_collection_params = namedtuple(.., ['upper', 'lower', 'collection_size'])`, though its not that important.